### PR TITLE
Share elasticsearch accounts

### PR DESCRIPTION
**Pull request by zanchi-r**
This PR adds the ability to share data between multiple elasticsearch indexes.
This is the first step to be able to implement account sharing between users

### CODE REVIEW

File: db/migration/0011_add_shared_accounts.sql [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+--   distributed under the License is distributed on an "AS IS" BASIS,
+--   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+--   See the License for the specific language governing permissions and
+--   limitations under the License.
+
+
+CREATE TABLE shared_account (
+  id                     INTEGER      NOT NULL AUTO_INCREMENT,
+  account_id             INTEGER      NOT NULL,
+  owner_id               INTEGER      NOT NULL,
```
Comment (thibautcornolti) [inline db/migration/0011_add_shared_accounts.sql]: No need to store `owner_id`

File: costs/tags/tags_keys.go [hunk 2]
```diff
@@ -70,27 +70,27 @@ func getTagsKeysWithParsedParams(ctx context.Context, params tagsKeysQueryParams
...
 	search.Aggregation("tags", elastic.NewTermsAggregation().Script(script))
 	res, err := search.Do(ctx)
 	if err != nil {
 		if elastic.IsNotFound(err) {
-			l.Warning("Query execution failed, ES index does not exists : " + index, err)
+			l.Warning("Query execution failed, ES index does not exists : "+index, err)
 			return nil, http.StatusOK, err
 		}
-		l.Error("Query execution failed : " + err.Error(), nil)
+		l.Error("Query execution failed : "+err.Error(), nil)
```
Comment (thibautcornolti) [inline costs/tags/tags_keys.go]: `l.Error("Query execution failed", err.Error())`

File: costs/tags/tags_keys.go [hunk 3]
```diff
@@ -70,27 +70,27 @@ func getTagsKeysWithParsedParams(ctx context.Context, params tagsKeysQueryParams
...
-	index := es.IndexNameForUser(user, "lineitems")
+	index := strings.Join(params.IndexList, ",")
 	script := elastic.NewScript("if (params._source.containsKey(\"tags\")) {  params._source[\"tags\"].keySet(); }")
 	search := client.Search().Index(index).Size(0).Query(query)
 	search.Aggregation("tags", elastic.NewTermsAggregation().Script(script))
 	res, err := search.Do(ctx)
 	if err != nil {
 		if elastic.IsNotFound(err) {
-			l.Warning("Query execution failed, ES index does not exists : " + index, err)
+			l.Warning("Query execution failed, ES index does not exists : "+index, err)
```
Comment (thibautcornolti) [inline costs/tags/tags_keys.go]: `l.Warning("Query execution failed, ES index does not exists", map[string]interface{}{"index": index, "error": err})`

File: costs/tags/tags_values.go [hunk 4]
```diff
@@ -90,27 +89,27 @@ func getTagsValuesWithParsedParams(ctx context.Context, params tagsValuesQueryPa
...
+	index := strings.Join(params.IndexList, ",")
 	search := client.Search().Index(index).Size(0).Query(query)
-	search.Aggregation("tags", elastic.NewTermsAggregation().Field("tags." + params.TagsKey[i]).
+	search.Aggregation("tags", elastic.NewTermsAggregation().Field("tags."+params.TagsKey[i]).
 		SubAggregation("cost", elastic.NewSumAggregation().Field("unblendedCost")))
 	res, err := search.Do(ctx)
 	if err != nil {
 		if elastic.IsNotFound(err) {
-			l.Warning("Query execution failed, ES index does not exists : " + index, err)
+			l.Warning("Query execution failed, ES index does not exists : "+index, err)
```
Comment (thibautcornolti) [inline costs/tags/tags_values.go]: Same as above

File: costs/tags/tags_values.go [hunk 5]
```diff
@@ -90,27 +89,27 @@ func getTagsValuesWithParsedParams(ctx context.Context, params tagsValuesQueryPa
...
 			return nil, http.StatusOK, err
 		}
-		l.Error("Query execution failed : " + err.Error(), nil)
+		l.Error("Query execution failed : "+err.Error(), nil)
```
Comment (thibautcornolti) [inline costs/tags/tags_values.go]: Same as above

File: es/accountsSharing.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+// AccountsAndIndexes stores the accounts and indexes
+type AccountsAndIndexes struct {
+	Accounts []string
+	Indexes  []string
+}
+
+// isAccountDuplicate returns true if the account already exists in the list of accounts
+// returns false otherwise
+func (ai *AccountsAndIndexes) isAccountDuplicate(account string) bool {
```
Comment (thibautcornolti) [inline es/accountsSharing.go:36]: `isAccountExists` would be better I guess

File: es/accountsSharing.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+		if entry == index {
+			return
+		}
+	}
+	ai.Indexes = append(ai.Indexes, index)
+}
+
+// GetAccountIdFromRoleArn takes a role arn string and returns the account ID
+// or an empty string if the role arn is not valid
+func GetAccountIdFromRoleArn(roleArn string) string {
```
Comment (thibautcornolti) [inline es/accountsSharing.go:66]: Add the checking of the length (12 for an account id).

**Review by thibautcornolti — DISMISSED** [review]

File: es/accountsSharing.go (hunk 7 above)
Comment (zanchi-r) [reply es/accountsSharing.go:66]: Not necessary, the account in DB should be formatted properly; if it is not there will be no data anyway

File: es/accountsSharing.go (hunk 6 above)
Comment (zanchi-r) [reply es/accountsSharing.go:36]: Duplicate is more explicit

File: db/schema.sql [hunk 8]
```diff
@@ -312,8 +311,8 @@ CREATE TABLE shared_account (
   id                     INTEGER      NOT NULL AUTO_INCREMENT,
   account_id             INTEGER      NOT NULL,
   user_id                INTEGER      NOT NULL,
-  user_permission        INTEGER   NOT NULL DEFAULT 0,
-  account_status         BOOL   NOT NULL DEFAULT 0,
+  user_permission        INTEGER  	  NOT NULL DEFAULT 0,
```
Comment (simonmeyerrr) [inline db/schema.sql]: there is a \t (not visible on Goland) between `INTEGRER` and `NOT NULL...` that make a bad indentation

File: db/schema.sql [hunk 9]
```diff
@@ -312,8 +311,8 @@ CREATE TABLE shared_account (
...
+  sharing_accepted         BOOL         NOT NULL DEFAULT 0,
```
Comment (simonmeyerrr) [inline db/schema.sql]: Bad indentation

File: costs/anomalies/anomalies_route.go [hunk 10]
```diff
@@ -67,6 +70,9 @@ func init() {
 func makeElasticSearchRequest(ctx context.Context, parsedParams AnomalyEsQueryParams, user users.User) (*elastic.SearchResult, int, error) {
 	l := jsonlog.LoggerFromContextOrDefault(ctx)
 	index := es.IndexNameForUser(user, "lineitems")
+	if len(parsedParams.IndexList) > 0 {
```
Comment (simonmeyerrr) [inline costs/anomalies/anomalies_route.go]: Why do you do this like this and not like every other functions that you modified ?

File: db/schema.sql [hunk 11]
```diff
@@ -277,7 +277,6 @@ CREATE TABLE emailed_anomaly (
 	CONSTRAINT PRIMARY KEY (id),
 	CONSTRAINT foreign_aws_account FOREIGN KEY (aws_account_id) REFERENCES aws_account(id) ON DELETE CASCADE
 );
-
```
Comment (simonmeyerrr) [inline db/schema.sql]: There many with and many withtout empty line, what should we do everywhere ?

File: db/schema.sql (hunk 8 above)
Comment (zanchi-r) [reply db/schema.sql]: fixed

File: db/schema.sql (hunk 9 above)
Comment (zanchi-r) [reply db/schema.sql]: fixed

File: costs/anomalies/anomalies_route.go (hunk 10 above)
Comment (zanchi-r) [reply costs/anomalies/anomalies_route.go]: I forgot to remove it, good catch!

File: db/schema.sql (hunk 11 above)
Comment (zanchi-r) [reply db/schema.sql]: I don't think we really car about that, but I added the line back

File: costs/anomalies/anomalies_route.go [hunk 12]
```diff
@@ -66,7 +69,7 @@ func init() {
 // with empty data
 func makeElasticSearchRequest(ctx context.Context, parsedParams AnomalyEsQueryParams, user users.User) (*elastic.SearchResult, int, error) {
 	l := jsonlog.LoggerFromContextOrDefault(ctx)
-	index := es.IndexNameForUser(user, "lineitems")
+	index = strings.Join(parsedParams.IndexList, ",")
```
Comment (Lolincolc) [inline costs/anomalies/anomalies_route.go]: You need to add ':' after index as index does not already exist

**Review by Lolincolc — DISMISSED** [review]

**Review by simonmeyerrr — APPROVED** [review]

**Pull request merged**